airsonic-main/src/main/java/org/airsonic/player/Application.java

```diff
@@ -162,6 +165,13 @@ public Filter noCacheFilter() {
     }
 
     private static SpringApplicationBuilder doConfigure(SpringApplicationBuilder application) {
+        // Handle HSQLDB database upgrades from 1.8 to 2.x before any beans are started.
+        application.application().addListeners((ApplicationListener<ApplicationPreparedEvent>) event -> {
+            if (event.getApplicationContext().getEnvironment().acceptsProfiles("legacy")) {
+                LegacyHsqlUtil.upgradeHsqldbDatabaseSafely();
+            }
```

Comment:
I _think_ this works, though you might want to check that this event listener isn't called everytime there is a context refresh. The bean definitions shouldn't matter so long as the context isn't refreshed.